salt/requisite.py

```diff
@@ -0,0 +1,412 @@
+                        )
+                    found = self.add_dependency(
+                        low, r_type, req_key, req_val)
+                    if not found:
+                        return ("Referenced state does not exist"
+                                f" for requisite [{r_type}: ({req_key}: {req_val})] in state"
+                                f" [{low['name']}] in SLS [{low.get('__sls__')}]")
+        return None
+
+    def aggregate_and_order_chunks(self, cap: int) -> list[LowChunk]:
+        """
+        Aggregate eligible nodes in the dependencies graph.
+
+        Return a list of the chunks in the sorted order in which the
+        chunks should be executed.
+        Nodes are eligible for aggregation if the state function in the
+        chunks match and aggregation is enabled in the configuration for
+        the state function.
+        :param cap: the maximum order value configured in the states
+        :return: the ordered chunks
+        """
+        dag: nx.MultiDiGraph = self.dag
+        # dict for tracking topo order and for mapping each node that
+        # was aggregated to the aggregated node that replaces it
+        topo_order = {}
+
+        max_group_size = 500
+        groups_by_type = defaultdict(list)
+
+        def _get_order(node):
+            chunk = dag.nodes[node].get("chunk", {})
+            chunk_order = self._get_chunk_order(cap, node)
+            return (chunk_order, "{0[state]}{0[name]}{0[fun]}".format(chunk))
+
+        # Iterate over the nodes in topological order to get the correct
+        # ordering which takes requisites into account
+        for i, node in enumerate(
+                nx.lexicographical_topological_sort(
+                    dag,
+                    key=_get_order)):
+            topo_order[node] = None
+            data = dag.nodes[node]
+            if not data.get('allow_aggregate'):
+                continue
+
+            node_type = data['state']
+            added = False
+            for i, group in enumerate(groups_by_type[node_type]):
+                if len(group) >= max_group_size:
+                    continue
+                # Check if the node can be reached from any node in the group
+                first_node = next(iter(group))
+                agg_node = topo_order.get(first_node)
+                # Since we are iterating in topological order we know
+                # that there is no path from the current node to the
+                # node in the group; so we only need to check the path
+                # from the group node to the current node
+                reachable = nx.has_path(dag, agg_node or first_node, node)
+                if not reachable:
+                    # If not, add the node to the group
+                    if agg_node is None:
+                        # there is now more than one node for this
+                        # group so aggregate them
+                        agg_node = f'__aggregate_{node_type}_{i}__'
```

Comment:
```suggestion
            for idx, group in enumerate(groups_by_type[node_type]):
                if len(group) >= max_group_size:
                    continue
                # Check if the node can be reached from any node in the group
                first_node = next(iter(group))
                agg_node = topo_order.get(first_node)
                # Since we are iterating in topological order we know
                # that there is no path from the current node to the
                # node in the group; so we only need to check the path
                # from the group node to the current node
                reachable = nx.has_path(dag, agg_node or first_node, node)
                if not reachable:
                    # If not, add the node to the group
                    if agg_node is None:
                        # there is now more than one node for this
                        # group so aggregate them
                        agg_node = f'__aggregate_{node_type}_{idx}__'
```